### PR TITLE
[Culling][Optimization] Fustum culling with temporal coherence

### PR DESCRIPTION
**Pull request by bigbike**
Let the simulator memorize the frustum plane index, which culls the object in the last frame.
In the current frame, it is likely the object will be rejected again by this plane. Thus, start the culling from this plane first.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/esp/gfx/RenderCamera.cpp [hunk 1]
```diff
@@ -15,6 +15,41 @@ namespace Mn = Magnum;
...
+};
+
+/**
+ * @brief do frustum culling with temporal coherence
+ * @param range, the axis-aligned bounding box
+ * @param frustum, the frustum
+ * @param frustumPlaneIndex, the frustum plane in last frame that culled the
+ * aabb (default: 0)
+ */
+CullingResult rangeFrustum(const Mn::Range3D& range,
```
Comment (erikwijmans) [inline src/esp/gfx/RenderCamera.cpp]: I'd prefer using an https://doc.magnum.graphics/corrade/classCorrade_1_1Containers_1_1Optional.html here.

Comment (bigbike) [reply src/esp/gfx/RenderCamera.cpp]: Can I ask why?

Comment (erikwijmans) [reply src/esp/gfx/RenderCamera.cpp]: Doesn't requiring introducing an additional type.

Comment (bigbike) [reply src/esp/gfx/RenderCamera.cpp]: I see. You were actually talking the return type. I thought you were referring the arguments.

That is a good idea.

Comment (erikwijmans) [reply src/esp/gfx/RenderCamera.cpp]: Ah, apologies for lack of clarity.

Comment (bigbike) [reply src/esp/gfx/RenderCamera.cpp]: Not a problem at all. Good advice, makes the code simpler. Will change.

**Comment by bigbike** [comment]
Dear reviewers,

Today, I designed an experiment to measure the performance gain from leveraging temporal coherence (denoted as TC). 

The test code can be found [here](https://github.com/facebookresearch/habitat-sim/compare/culling-test_op2...op2-eval).

To evaluate the performance, I designed 2 metrics:
- ratio between the number of culled objects that benefit from TC and the total number of culled objects;
- ratio between the number of frustum plane tests when TC is turned ON / OFF;

The result is as follows:
```
 ========================= Performance ========================
 256 x 256, total time 7.45 s, frame time 1.863 ms (536.8 FPS)
 ==============================================================
###### The efficiency of temporal coherence ######
culled objects that benefit from temporal coherency / total culled objects: 1096767 / 1115010, ratio: 98.36%
plane tests with and without temporal coherence enabled: 12560284283 / 14545762971, SAVE: 13.65%
```

**Conclusions:**
- Temporal coherence almost exists in every culled object (**~99%**);
- With temporal coherence heuristic, it roughly saves **10%** visibility tests;



The command line I used:
`python examples/example.py --scene ~/models/basis-mp3d/dhjEzFoUFzH/dhjEzFoUFzH.glb --width 256 --height 256 --max_frames 4000 --silent --seed 5`

**Comment by mosra** [comment]
This is good! :+1:

**Pull request merged**